lib/collection/src/discovery.rs

```diff
@@ -0,0 +1,234 @@
+    });
+
+    // Fetch all referenced vectors
+    let referenced_vectors = id_references
+        .fetch_vectors(collection, read_consistency, collection_by_name)
+        .await?;
+
+    // Create core search requests
+    let mut core_searches = Vec::with_capacity(request_batch.searches.len());
+    for request in &request_batch.searches {
+        let lookup_collection_name = request.lookup_from.as_ref().map(|x| &x.collection);
+
+        let lookup_vector_name = get_vector_name(request);
+
+        // Check we actually fetched all referenced vectors in this request
+        let referenced_ids = iterate_examples(request)
+            .filter_map(RecommendExample::as_point_id)
+            .collect_vec();
+
+        referenced_ids.iter().try_for_each(|&point_id| {
+            if !referenced_vectors.contains_key(&PointRef {
+                collection_name: lookup_collection_name,
+                point_id,
+            }) {
+                return Err(CollectionError::PointNotFound {
+                    missed_point_id: point_id,
+                });
+            }
+            Ok(())
+        })?;
+
+        let target = convert_to_vectors(
+            request.target.iter(),
+            &referenced_vectors,
+            &lookup_vector_name,
+            lookup_collection_name,
+        )
+        .next()
+        .cloned();
+
+        let context_pairs = request
+            .context_pairs
+            .iter()
+            .flatten()
+            .map(|pair| {
+                let mut vector_pair = convert_to_vectors(
+                    std::iter::once(&pair.0).chain(std::iter::once(&pair.1)),
+                    &referenced_vectors,
+                    &lookup_vector_name,
+                    lookup_collection_name,
+                )
+                .cloned();
+
+                ContextPair {
+                    // SAFETY: we know there are two elements in the iterator
+                    positive: vector_pair.next().unwrap(),
+                    negative: vector_pair.next().unwrap(),
+                }
+            })
+            .collect_vec();
+
+        let query: QueryEnum = match (target, context_pairs) {
+            // Only target => Nearest
+            (Some(target), pairs) if pairs.is_empty() => QueryEnum::Nearest(
+                NamedVector {
+                    name: lookup_vector_name,
+                    vector: target,
+                }
+                .into(),
+            ),
+
+            // Target and pairs => Discovery
+            (Some(target), pairs) => QueryEnum::Discover(NamedQuery {
+                query: DiscoveryQuery::new(target, pairs),
+                using: Some(lookup_vector_name),
+            }),
+
+            // Only pairs => Context
+            (None, pairs) => QueryEnum::Context(NamedQuery {
+                query: ContextQuery::new(pairs),
+                using: Some(lookup_vector_name),
+            }),
+        };
```

Comment:
It makes a bit more sense to me if it is ordered like this:

```suggestion
        let query: QueryEnum = match (target, context_pairs) {
            // Target and pairs => Discovery
            (Some(target), pairs) if !pairs.is_empty() => QueryEnum::Discover(NamedQuery {
                query: DiscoveryQuery::new(target, pairs),
                using: Some(lookup_vector_name),
            }),

            // Only target => Nearest
            (Some(target), _pairs) => QueryEnum::Nearest(
                NamedVector {
                    name: lookup_vector_name,
                    vector: target,
                }
                .into(),
            ),

            // Only pairs => Context
            (None, pairs) => QueryEnum::Context(NamedQuery {
                query: ContextQuery::new(pairs),
                using: Some(lookup_vector_name),
            }),
        };
```

- Discovery
- Nearest
- Context